cEP-0022.md

```diff
@@ -0,0 +1,162 @@
+
+This document describes what is the **Green Mode** for `coala-quickstart` and how
+will this be implemented.
+
+This cEP describes new type of config files which will be generated by
+`coala-quickstart`, how will they be different from the ones currently being
+created, how is this helpful and the action-plan for adding this feature.
+
+# Introduction
+
+Currently `coala-quickstart` generates config files which point out too many
+errors in the code-base of the project. It also asks too many questions
+while generating the config file. The config files are not that specific to
+the project and the errors pointed out by coala using generated config files
+are generally undesirable for the project owners.
+
+Project owners need a configuration file which *agrees* with their codestyle, is
+as specific as possible i.e. can point out maximum errors with least amount
+of manual edits in the future, for future inconsistencies in commits.
+Many organizations don't follow specific code styles like `PEP8`, instead they
+follow their own standards. We need to adapt to those standards with least
+user interaction and produce config files which generate absolutely no errors
+in their codebase. This will lead to easier adoption of coala across communities
+as it is going to result in an easy mergeable Pull Request regarding
+*adding `.coafile` and CI to enforce it* in the other orgs for example
+<https://github.com/processing/processing/pull/4781>.
+
+# Implementation
+
+This project will be implemented in two phases:
+
+- The Brute Force way
+- Further optimization on the Brute Force
+
+1. The first step will consist of preparing a dict of all the values that can
+   be attained by the bears' settings, may it be necessary or optional as we need
+   to create *green* config files that are as specific as possible. Failure to do
+   so will lead to dropping of most **optional** bear settings, thus will lead to
+   coala not detecting many inconsistencies, and the developers of the project will
+   have to do the work of learning about the bears and their settings and then
+   modifying the `.coafile` manually, while not able to guess the correct
+   **necessary** bear settings will lead to dropping of bears themselves.
+
+**The following steps mainly describe the brute force way**
+
+2. Certain thresholds may be defined for an org based on several factors
+   for the number of inconsistencies detected to be considered mistakes however
+   these shall be stored for later as the first and foremost goal of this project
+   is to generate completely *"green"* config files.
+
+3. coala will be ran over and over again for the entire project with all
+   instances of bears with all the combinations of settings running in
+   parallel, as soon as instances of a particular setting is done and we get
+   a value for the setting for which no errors were generated, it goes directly
+   into the `.coafile`, if no value for an optional setting matches, the setting will
```

Comment:
>matches

What does a match mean? 0 errors? What about a scenario where most of the code base uses spaces, except for two or three lines using tabs. In this case, we want coala to identify this as an issue ("correct setting is spaces, these two lines are in violation") instead of dropping the bear/setting altogether.

Maybe a percentage based heurestic would work. Of the N files passed to a bear, if M are okay, we consider the okay-ness of a bear/setting to be `M/N`. This is doubly useful as small settings would automatically get thrown out: if there are only 3 files, and one of them is found to be in violation, the okayness of the bear/setting is ~66%. We can consider this to be below a threshold (heurestic, obviously; needs a lot of trial and error on real projects to come up with the right number), and *then* discard the bear.

Absolute number of errors (`N-M`) might be misleading in large projects.

Reply:
that was my initial request (i overdid it while writing the proposal) but apparently we want completely green config files which don't point out even a single error in the project for starters. after that we can improve upon it and request for some modifications

Reply:
While dropping the bears due to just small percent on inconsistencies, we can issue a warning to the user.